File: velocity-sdk-python/src/velocity/connection.py
"""gRPC connection management for V.E.L.O.C.I.T.Y.-WorkFlow Python SDK"""

from typing import Any, Dict, Optional
from .types import WorkflowExecution, WorkflowOptions, HistoryEvent


class Connection:
    """Manages gRPC connection to the V.E.L.O.C.I.T.Y.-WorkFlow server"""

    def __init__(self, host_port: str, tls: bool = False):
        self.host_port = host_port
        self.tls = tls
        self.channel = None
        self._connected = False

    def connect(self) -> None:
        """Establish connection to the server"""
        if self._connected:
            return

        import grpc

        if self.tls:
            credentials = grpc.ssl_channel_credentials()
            self.channel = grpc.secure_channel(self.host_port, credentials)
        else:
            self.channel = grpc.insecure_channel(self.host_port)

        self._connected = True

    def close(self) -> None:
        """Close the connection"""
        if self.channel:
            self.channel.close()
            self._connected = False

    def is_connected(self) -> bool:
        """Check if connected"""
        return self._connected and self.channel is not None

    def start_workflow(self, namespace: str, options: WorkflowOptions) -> WorkflowExecution:
        """Start a new workflow execution"""
        # In a real implementation, this would call the gRPC client
        import time
        from .types import WorkflowStatus
        return WorkflowExecution(
            workflow_id=options.workflow_id,
            run_id=f"run-{options.workflow_id}-{int(time.time() * 1000)}",
            workflow_type=options.workflow_type,
            task_queue=options.task_queue,
            status=WorkflowStatus.RUNNING,
            started_at=int(time.time() * 1000),
        )

    def signal_workflow(self, namespace: str, workflow_id: str, signal_name: str, input: Any) -> None:
        """Signal a running workflow"""
        pass

    def query_workflow(self, namespace: str, workflow_id: str, query_type: str, input: Any) -> Any:
        """Query a running workflow"""
        return None

    def terminate_workflow(self, namespace: str, workflow_id: str, reason: str) -> None:
        """Terminate a running workflow"""
        pass

    def cancel_workflow(self, namespace: str, workflow_id: str) -> None:
        """Cancel a running workflow"""
        pass

    def describe_workflow(self, namespace: str, workflow_id: str) -> Optional[WorkflowExecution]:
        """Get workflow execution details"""
        return None

    def get_workflow_history(self, namespace: str, workflow_id: str) -> list:
        """Get workflow execution history"""
        return []

File: velocity-sdk-typescript/src/connection.ts
/**
 * HTTP Connection to Velocity server
 */

export interface ConnectionOptions {
  address: string;
  tls?: boolean;
  headers?: Record<string, string>;
}

export class Connection {
  private baseUrl: string;
  private headers: Record<string, string>;

  constructor(options: ConnectionOptions) {
    let address = options.address;
    if (!address.startsWith('http://') && !address.startsWith('https://')) {
      const scheme = options.tls ? 'https' : 'http';
      address = `${scheme}://${address}`;
    }
    this.baseUrl = address.replace(/\/+$/, '');
    this.headers = {
      'Content-Type': 'application/json',
      ...options.headers,
    };
  }

  private async request<T = any>(method: string, path: string, body?: any): Promise<T> {
    const url = `${this.baseUrl}${path}`;
    const resp = await fetch(url, {
      method,
      headers: this.headers,
      body: body ? JSON.stringify(body) : undefined,
    });

    const text = await resp.text();
    if (!resp.ok) {
      throw new Error(`HTTP ${resp.status}: ${text}`);
    }

    if (text.length === 0) return undefined as any;
    try {
      return JSON.parse(text) as T;
    } catch {
      return text as any;
    }
  }

  async startWorkflow(params: {
    namespace: string;
    workflowId: string;
    workflowType: string;
    taskQueue: string;
    input?: any;
  }): Promise<{ workflowId: string; runId: string }> {
    return this.request('POST', '/api/workflows', params);
  }

  async signalWorkflow(params: {
    namespace: string;
    workflowId: string;
    signalName: string;
    input?: any;
  }): Promise<void> {
    await this.request('POST', `/api/workflows/${params.workflowId}/signal`, params);
  }

  async queryWorkflow(params: {
    namespace: string;
    workflowId: string;
    queryType: string;
    input?: any;
  }): Promise<any> {
    return this.request('POST', `/api/workflows/${params.workflowId}/query`, params);
  }

  async terminateWorkflow(params: {
    namespace: string;
    workflowId: string;
    reason?: string;
  }): Promise<void> {
    await this.request('POST', `/api/workflows/${params.workflowId}/terminate`, params);
  }

  async cancelWorkflow(params: {
    namespace: string;
    workflowId: string;
  }): Promise<void> {
    await this.request('POST', `/api/workflows/${params.workflowId}/cancel`, params);
  }

  async describeWorkflow(params: {
    namespace: string;
    workflowId: string;
  }): Promise<any> {
    return this.request('GET', `/api/workflows/${params.workflowId}`);
  }

  async getWorkflowHistory(params: {
    namespace: string;
    workflowId: string;
  }): Promise<any[]> {
    try {
      return await this.request('GET', `/api/workflows/${params.workflowId}/history`);
    } catch {
      return [];
    }
  }

  async listWorkflows(namespace: string): Promise<any[]> {
    try {
      const resp = await this.request<{ workflows: any[] }>('GET', '/api/workflows');
      return resp?.workflows || [];
    } catch {
      return [];
    }
  }

  async healthCheck(): Promise<boolean> {
    try {
      const resp = await this.request<{ status: string }>('GET', '/api/health');
      return resp?.status === 'healthy' || resp?.status === 'ok';
    } catch {
      return false;
    }
  }

  // Polling methods (for worker task processing)
  async pollWorkflowTaskQueue(_params: {
    namespace: string;
    taskQueue: string;
  }): Promise<any> {
    // In HTTP mode, workflow tasks come from the REST API polling
    return null;
  }

  async pollActivityTaskQueue(_params: {
    namespace: string;
    taskQueue: string;
  }): Promise<any> {
    // In HTTP mode, activity tasks come from the REST API polling
    return null;
  }

  async respondWorkflowTaskCompleted(_params: {
    taskToken: string;
    commands: any[];
  }): Promise<void> {
    // In HTTP mode, workflow task responses are sent via the REST API
  }

  async respondActivityTaskCompleted(_params: {
    taskToken: string;
    result?: any;
  }): Promise<void> {
    // In HTTP mode, activity task responses are sent via the REST API
  }

  async respondActivityTaskFailed(_params: {
    taskToken: string;
    failure: string;
  }): Promise<void> {
    // In HTTP mode, activity task failures are reported via the REST API
  }

  close(): void {
    // No-op for HTTP connections
  }
}
